SoftwareGuide/Latex/DevelopmentGuidelines/CreateAModule.tex

```diff
@@ -929,6 +930,151 @@ \subsubsection{Wrapping Tests}
+    \code{CMAKE\_BUILD\_TYPE:STRING="RelWithDebInfo"} and \code{ITK\_WRAP\_PYTHON:BOOL="On"}.
+    Note that the "RelWithDebInfo" build type is strongly encouraged over a "Debug"
+    build as the former will build against a standard Python distribution.
+    See Section~\ref{sec:UsingCMakeForConfiguringAndBuildingITK} for a detailed
+    explanation of how to build ITK locally with CMake.
+  \item A Python virtual environment must be appropriately configured with
+    \code{WrapITK.pth} so that the ITK debug build can be loaded.
+    See Section~\ref{sec:Wrapping} for further explanation.
+  \item ITK modules to be debugged must be loaded in a new Python session initialized
+    from the given virtual environment. Given that ITK Python uses lazy loading, it is
+    pragmatic to use \code{itk.force\_load()} to ensure that all possible debug symbols
+    are made available. The \code{os} module can be used to identify the PID of the given session.
+
+\begin{minted}[baselinestretch=1,fontsize=\footnotesize,linenos=false,bgcolor=ltgray]{bash}
+(venv-itk) > python
+>>> import itk
+>>> itk.force_load()
+>>> import os
+>>> os.getpid()
+99999
+\end{minted}
+
+  \item The Python session may be attached to a debugger using the returned PID.
+    \begin{itemize}
+      \item On a Windows operating system Microsoft Visual Studio 2019 or a similar platform
+        can be used for attaching to a running process for debugging. Select "Attach To Process"
+        from the Debug menu, choose "Native" code, and then search for the process PID. If
+        debug symbols were generated correctly then ITK modules will appear under the list of
+        loaded modules.
+      \item On a Linux operating system the GNU Project Debugger \code{gdb} can be used
+        for attaching to a running Python process, reading symbol files, and setting breakpoints.
+        The following example attaches to a running process and sets a breakpoint inside
+        an \code{itk.Image} Python object.
+
+\begin{minted}[baselinestretch=1,fontsize=\footnotesize,linenos=false,bgcolor=ltgray]{bash}
+(venv-itk) > sudo gdb
+(gdb) > attach 99999
+Reading symbols from /path/to/ITK-build/Wrapping/Generators/Python/itk/_ITKPyBasePython.so...
+Reading symbols from /path/to/ITK-build/Wrapping/Generators/Python/itk/_ITKCommonPython.so...
+(gdb) > break /path/to/ITK-build/Wrapping/Modules/ITKCommon/itkImagePython.cpp:<ln>
+Breakpoint 1 at ...:
+file /path/to/ITK-build/Wrapping/Modules/ITKCommon/itkImagePython.cpp, line <ln>.
+(gdb) > break /path/to/ITK-source/Core/Common/include/itkImage.hxx:<ln>
+Breakpoint 2 at ...:
+/path/to/ITK-source/Modules/Core/Common/include/itkImage.hxx:<ln>
+(gdb) > c
+
+... continue in Python session until breakpoint is hit ...
+\end{minted}
+
+    \end{itemize}
+
+\end{enumerate}
+
+With these steps completed the respective debugger can be used to set breakpoints
+and step through C++ source code for a respective ITK Python execution.
+
```

Comment:
We could also add a section here on how to get a backtrace when a Python process crashes with a segfault. With Linux / gdb, this is setting `ulimit -c` and invoking `gdb` as described here:

https://pythondev.readthedocs.io/debug_tools.html#create-a-core-dump-file
https://nick-otter.github.io/linux/kernel/memory/2020/11/25/how-to-debug-crashes-and-segfaults/

I am not sure of the process on Windows / macOS, but that could be added if you can find it.

Reply:
Great idea, I'll look into the Windows side. I'm limited as to how much I can investigate on macOS, but I imagine the process is very similar to Linux?

Reply:
Great, thanks! @dzenanz or @brad-t-moore may know the incantations for Windows.

macOS is similar, although the standard debugger is now `lldb`. You can ssh into `misty` if you need a system to experiment on.

Reply:
It looks like Visual Studio makes the program call stack, locals, etc available on breakpoints or crashes, and the user can manually create and then reload dumps as desired. Process is described here, I will add a summary and URL reference: [https://docs.microsoft.com/en-us/visualstudio/debugger/using-dump-files?view=vs-2022](https://docs.microsoft.com/en-us/visualstudio/debugger/using-dump-files?view=vs-2022)

Reply:
As a status update here, I'm having trouble attaching LLDB to a Python process on `misty`:
```sh
error: process exited with status -1 (this is a non-interactive debug session, cannot get permission to debug processes.)
```

This occurs for an interactive LLDB session for both interactive and non-interactive Python sessions. LLDB seems to attach to other processes just fine. Running with elevated permissions does not resolve the issue, and I haven't been able to find much discussion online. The machine is in developer mode. Would welcome discussion from macOS users who may have seen this before  😃